sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala

```diff
@@ -176,7 +176,13 @@ object SQLExecution extends Logging {
             shuffleIds.foreach { shuffleId =>
               queryExecution.shuffleCleanupMode match {
                 case RemoveShuffleFiles =>
-                  SparkEnv.get.shuffleManager.unregisterShuffle(shuffleId)
+                  // Do not unregister the shuffle on MapOutputTracker here to trigger stage retry.
+                  // Otherwise, downstream tasks will fail with MetadataFetchFailedException.
+                  if (sc.isLocal) {
+                    SparkEnv.get.shuffleManager.unregisterShuffle(shuffleId)
+                  } else {
+                    sc.shuffleDriverComponents.removeShuffle(shuffleId, false)
```

Comment:
does this work for local mode?

Reply:
Yes `ShuffleDriverComponents.removeShuffle` should work for both local an non-local mode. Will update the code here.